src/lib.rs

```diff
@@ -235,6 +232,108 @@ fn get_main_camera_position(
     (origin.as_tile_index(), radius)
 }
 
+#[derive(Component)]
+struct Compass;
+
+#[derive(SystemParam)]
+struct Player<'w, 's> {
+    xr_pos: Query<
+        'w,
+        's,
+        (&'static Transform, &'static GalacticGrid),
+        (With<OpenXRTrackingRoot>, Without<FlyCam>, Without<Compass>),
+    >,
+    flycam_pos: Query<
+        'w,
+        's,
+        (&'static Transform, &'static GalacticGrid),
+        (With<FlyCam>, Without<OpenXRTrackingRoot>, Without<Compass>),
+    >,
+    space: Res<'w, FloatingOriginSettings>,
+}
+
+struct PlayerPosition<'a> {
+    transform: Transform,
+    grid: GalacticGrid,
+    space: &'a FloatingOriginSettings,
+}
+
+impl PlayerPosition<'_> {
+    pub fn pos(&self) -> DVec3 {
+        self.space.grid_position_double(&self.grid, &self.transform)
+    }
+    pub fn directions(&self) -> Directions {
+        let up = self.pos().normalize().as_vec3();
+        let west = Vec3::Z.cross(up);
+        let north = up.cross(west);
+        Directions { up, north, west }
+    }
+}
+
+struct Directions {
+    up: Vec3,
+    north: Vec3,
+    west: Vec3,
+}
+
+impl<'w, 's> Player<'w, 's> {
+    pub fn pos(&self) -> PlayerPosition<'_> {
+        let (&transform, &grid) = if let Ok(xr_pos) = self.xr_pos.get_single() {
+            xr_pos
+        } else {
+            self.flycam_pos.single()
+        };
+        PlayerPosition {
+            transform,
+            grid,
+            space: &self.space,
+        }
+    }
+}
+
+fn reposition_compass(
+    mut compass: Query<
+        (&mut Transform, &mut GalacticGrid),
+        (With<Compass>, Without<FlyCam>, Without<OpenXRTrackingRoot>),
+    >,
+    mut commands: Commands,
+    mut meshes: ResMut<Assets<Mesh>>,
+    mut materials: ResMut<Assets<StandardMaterial>>,
+    server: Res<AssetServer>,
+    player: Player,
+) {
+    if let Ok((mut pos, mut grid)) = compass.get_single_mut() {
+        let player = player.pos();
+        let directions = player.directions();
+        pos.translation = player.transform.translation - directions.up * 5.;
+        *grid = player.grid;
+        pos.look_to(directions.north, directions.up)
```

Comment:
@DerKarlos I added a helper that wraps all the annoying math in a nice function and gives you a struct with a `north`, `up`, and `west` vector (all orthogonal to each other).